### PR TITLE
Add Vultr provider

### PR DESCRIPTION
**Pull request by Xinayder**
Related issue: #20 

I've used the existing DigitalOcean provider as a base for this one and it was easy to make it work. There's a catch, however, that the content of the DNS record must be enclosed in quotes in Vultr, so I don't know if this is an issue for libdns.

### CODE REVIEW

File: vultr/client.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/vultr/govultr"
+	"github.com/libdns/libdns"
+)
+
+type Client struct {
```
Comment (mholt) [inline vultr/client.go:13]: This could probably be un-exported, right?

File: vultr/provider.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	"strings"
+	"time"
+
+	"github.com/libdns/libdns"
+)
+
+// Provider implements the libdns interfaces for Vultr
+// Adapted from libdns/digitalocean to work with the Vultr API
+type Provider struct {
+	Client
```
Comment (mholt) [inline vultr/provider.go:14]: Ah, maybe find a way to unexport this, since there's nothing the user can configure on this.

**Review by mholt — CHANGES_REQUESTED** [review]
Sorry it took me so long to get around to this! Looks good to me, I just noticed a couple things. Nice job!

**Comment by mholt** [comment]
I'll create the repo and give you access to push to it. When you've finished making any revisions, feel free to push your code to that repo.

File: vultr/client.go (hunk 1 above)
Comment (Xinayder) [reply vultr/client.go:13]: I think so, it's only used internally by the package. I will test if it works by unexporting this variable.

**Comment by Xinayder** [comment]
Thanks, I already pushed the corrected code to the new repository.

> A lot of people who use libdns use it through Caddy. If you're interested, would you like to create the Vultr DNS provider module for Caddy too?

It's funny, I had just created a repository to implement the Vultr DNS provider. I copied everything from the Cloudflare one, now I just need to test it to check if it works. It's mostly done: https://github.com/Xinayder/caddy-dns-vultr

**Comment by mholt** [comment]
Oh, yeah, that's true. Good catch, make sure it's updated in go.mod too.

**Comment by Xinayder** [comment]
I have to rewrite the Vultr client because they released a new version last week, and it solves some problems I had (e.g. when you created a record, the API didn't return its ID).

**Comment by Xinayder** [comment]
I've updated the `libdns/vultr` client to use Vultr API v2 and it works:

```
2020/11/25 15:48:16.719	INFO	http	enabling automatic TLS certificate management	{"domains": ["test.xinayder.me"]}
2020/11/25 15:48:16.719	INFO	tls	cleaned up storage units
2020/11/25 15:48:16.719	INFO	autosaved config	{"file": "/home/alex/.config/caddy/autosave.json"}
2020/11/25 15:48:16.719	INFO	serving initial configuration
2020/11/25 15:48:16.727	INFO	tls.obtain	acquiring lock	{"identifier": "test.xinayder.me"}
2020/11/25 15:48:16.727	INFO	tls.obtain	lock acquired	{"identifier": "test.xinayder.me"}
2020/11/25 15:48:16.727	INFO	tls.issuance.acme	waiting on internal rate limiter	{"identifiers": ["test.xinayder.me"]}
2020/11/25 15:48:16.727	INFO	tls.issuance.acme	done waiting on internal rate limiter	{"identifiers": ["test.xinayder.me"]}
2020/11/25 15:48:18.032	INFO	tls.issuance.acme.acme_client	trying to solve challenge	{"identifier": "test.xinayder.me", "challenge_type": "dns-01", "ca": "https://acme-staging-v02.api.letsencrypt.org/directory"}
2020/11/25 15:50:20.462	ERROR	tls.obtain	will retry	{"error": "[test.xinayder.me] Obtain: [test.xinayder.me] solving challenges: waiting for solver *certmagic.DNS01Solver to be ready: timed out waiting for record to fully propagate; verify DNS provider configuration is correct - last error: <nil> (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/16781389/191527555) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)", "attempt": 1, "retrying_in": 60, "elapsed": 123.734936439, "max_duration": 2592000}
2020/11/25 15:51:21.519	INFO	tls.issuance.acme.acme_client	trying to solve challenge	{"identifier": "test.xinayder.me", "challenge_type": "dns-01", "ca": "https://acme-staging-v02.api.letsencrypt.org/directory"}
2020/11/25 15:51:53.969	INFO	tls.issuance.acme.acme_client	validations succeeded; finalizing order	{"order": "https://acme-staging-v02.api.letsencrypt.org/acme/order/16781389/191528784"}
2020/11/25 15:51:54.718	INFO	tls.issuance.acme.acme_client	successfully downloaded available certificate chains	{"count": 2, "first_url": "https://acme-staging-v02.api.letsencrypt.org/acme/cert/fa7376110b00330e8e24a96ba0873f88cca9"}
2020/11/25 15:51:54.718	INFO	tls.obtain	certificate obtained successfully	{"identifier": "test.xinayder.me"}
2020/11/25 15:51:54.718	INFO	tls.obtain	releasing lock	{"identifier": "test.xinayder.me"}
```

The next thing to do is adding pagination to `getDNSEntries`.

**Comment by mholt** [comment]
That `timed out waiting for record to fully propagate;` error is interesting... (probably not a bug in vultr implementation, just saying that I wonder how common this is)

**Comment by Xinayder** [comment]
> Although that `timed out waiting for record to fully propagate;` error is interesting... (probably not a bug in vultr implementation, just saying that I wonder how common this is)

I think it could be a problem with my network, but even switching to 4G took some time to validate.

**Comment by mholt** [comment]
I'm skeptical it's the network. I feel like the lookups should be finding it pretty quickly.

**Comment by Xinayder** [comment]
I have implemented pagination support in the method that lists the DNS entries: https://github.com/libdns/vultr/commit/a8fa24426be88768b2f53ba2098b03812a0462f0

I also have a question, is it a problem if every time I run `go get -u github.com/libdns/vultr` it adds indirect dependencies to `caddy-dns/vultr`?

**Comment by mholt** [comment]
> I also have a question, is it a problem if every time I run go get -u github.com/libdns/vultr it adds indirect dependencies to caddy-dns/vultr?

libdns should have no Caddy dependencies, so if that does not go away after `go mod tidy` then something is wrong, I think.

> EDIT: I tested it again and it had the same error when trying to solve the ACME challenge. It's weird because after it fails, it retries a minute later and it succeeds.

Try extending the timeout and see if it succeeds on the first try?

**Comment by Xinayder** [comment]
> Try extending the timeout and see if it succeeds on the first try?

How do I extend the timeout in the Caddyfile?

**Comment by mholt** [comment]
I'm not sure if it's configurable via the Caddyfile (yet) -- you'll probably have to modify the code.

**Review by mholt — APPROVED** [review]

**Comment by mholt** [comment]
Btw, I think that error is _probably_ orthogonal to your provider implementation so it doesn't need to block your uploading it to the repo. So I've approved this PR and will close it now. Still, if you get a handle on that timeout with the check, I'd love to know what you discover!

**Comment by Xinayder** [comment]
I think it might be the way my domain is setup that's causing this delay in solving the challenge.

I bought the domain on Namecheap and I use Vultr to manage its DNS zones because I can't use the Namecheap API until I've spent $50 or have lots of domains.

I just deployed a custom `caddy` binary with the new DNS provider and it worked, except it took about 7 minutes to get my services up and running. I issued a wildcard certificate and for some reason it downloaded the certificate twice (maybe it re-downloaded the certificate for my base domain?). It couldn't solve the certificate in 2 minutes, and it waited more time for the DNS to propagate, but it only failed 2 minutes after the DNS zone was added, the second time it attempted to solve the challenge it didn't fail, just took some time.

**Comment by mholt** [comment]
Do you have the logs for that? It shouldn't download certificates twice... never heard of that before, in fact. That whole thing really does sound strange. Does the problem happen if you use [lego](https://github.com/go-acme/lego)'s CLI? (You don't _have_ to go to the trouble to find out, I'm really just curious if this is a universal problem with that provider or if there's something wrong with my DNS challenge code.)

**Comment by Xinayder** [comment]
Sorry for taking long to reply, but here are the logs:

```
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4676633,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["*.aoalmeida.com","aoalmeida.com"]}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4805577,"logger":"tls.obtain","msg":"acquiring lock","identifier":"*.aoalmeida.com"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4816382,"logger":"tls.obtain","msg":"lock acquired","identifier":"*.aoalmeida.com"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4821758,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc00021b880"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.48657,"msg":"autosaved config","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4867744,"msg":"serving initial configuration"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4892125,"logger":"tls","msg":"cleaned up storage units"}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.490673,"logger":"tls.issuance.acme","msg":"waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:45:37 localhost caddy[26785]: {"level":"info","ts":1606599937.4909208,"logger":"tls.issuance.acme","msg":"done waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:45:38 localhost caddy[26785]: {"level":"info","ts":1606599938.2134566,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"*.aoalmeida.com","challenge_type":"dns-01","ca":"https://acme-v02.api.letsencrypt.org/directory"}
Nov 28 21:47:39 localhost caddy[26785]: {"level":"error","ts":1606600059.1309686,"logger":"tls.obtain","msg":"will retry","error":"[*.aoalmeida.com] Obtain: [*.aoalmeida.com] solving challenges: waiting for solver *certmagic.DNS01Solver to be ready: timed out waiting for record to fully propagate; verify DNS provider configuration is correct - last error: <nil> (order=https://acme-v02.api.letsencrypt.org/acme/order/101130370/6460603033) (ca=https://acme-v02.api.letsencrypt.org/directory)","attempt":1,"retrying_in":60,"elapsed":121.649082152,"max_duration":2592000}
Nov 28 21:48:39 localhost caddy[26785]: {"level":"info","ts":1606600119.6782331,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"*.aoalmeida.com","challenge_type":"dns-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Nov 28 21:50:20 localhost caddy[26785]: {"level":"info","ts":1606600220.7814963,"logger":"tls.issuance.acme.acme_client","msg":"validations succeeded; finalizing order","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16420384/193483835"}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.2520313,"logger":"tls.issuance.acme.acme_client","msg":"successfully downloaded available certificate chains","count":2,"first_url":"https://acme-staging-v02.api.letsencrypt.org/acme/cert/fa28c5dadb293c589c96f11af1de9ef671b7"}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.253891,"logger":"tls.issuance.acme","msg":"waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.254234,"logger":"tls.issuance.acme","msg":"done waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.736078,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"*.aoalmeida.com","challenge_type":"dns-01","ca":"https://acme-v02.api.letsencrypt.org/directory"}
Nov 28 21:52:22 localhost caddy[26785]: {"level":"info","ts":1606600342.8160934,"logger":"tls.issuance.acme.acme_client","msg":"validations succeeded; finalizing order","order":"https://acme-v02.api.letsencrypt.org/acme/order/101130370/6460658836"}
Nov 28 21:52:23 localhost caddy[26785]: {"level":"info","ts":1606600343.3013282,"logger":"tls.issuance.acme.acme_client","msg":"successfully downloaded available certificate chains","count":2,"first_url":"https://acme-v02.api.letsencrypt.org/acme/cert/03194c90cfbc26da8dfff28acdab340aabfb"}
Nov 28 21:52:23 localhost caddy[26785]: {"level":"info","ts":1606600343.302797,"logger":"tls.obtain","msg":"certificate obtained successfully","identifier":"*.aoalmeida.com"}
Nov 28 21:52:23 localhost caddy[26785]: {"level":"info","ts":1606600343.3030186,"logger":"tls.obtain","msg":"releasing lock","identifier":"*.aoalmeida.com"}
```

I used another domain that is also hosted on Vultr.

**Comment by mholt** [comment]
Thanks! Well... I'm glad the retry gets it, I still wonder if/why it doesn't happen with lego, and if not, what is different. But that's probably a different issue.

**Comment by Xinayder** [comment]
I find it weird that it issues 2 certificates and downloads both, instead of issuing a single one. I don't know how Caddy internals work but doesn't the following line indicate that there are 2 certificates to be downloaded?

```
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.2520313,"logger":"tls.issuance.acme.acme_client","msg":"successfully downloaded available certificate chains","count":2,"first_url":"https://acme-staging-v02.api.letsencrypt.org/acme/cert/fa28c5dadb293c589c96f11af1de9ef671b7"}
```

Maybe it's issuing one certificate per attempt, but I just checked and it didn't renew the certificate for the root domain, maybe it was attempting to do so?

Also, I've noticed that Caddy had 2 different challenges added to my DNS zones.

**Comment by mholt** [comment]
@Xinayder Ah, I see what you mean, this?

>  `"msg":"successfully downloaded available certificate chains","count":2`

Fortunately, that's not a problem. The CA offered 2 chains in this case. ACMEz downloads all available chains by default, then CertMagic just keeps the first one (unless an alternate is configured).

> Also, I've noticed that Caddy had 2 different challenges added to my DNS zones.

This is probably because a DNS challenge for `*.example.com` and `example.com` use the same ACME challenge subdomain. In fact, maybe if it's trying to do both at once is why the first one fails... maybe it's just a matter of not overlapping challenges for domains and their associated wildcard variant.

**Comment by Xinayder** [comment]
> This is probably because a DNS challenge for `*.example.com` and `example.com` use the same ACME challenge subdomain. In fact, maybe if it's trying to do both at once is why the first one fails... maybe it's just a matter of not overlapping challenges for domains and their associated wildcard variant.

Maybe it's not actually trying to validate both domains, because as seen in the log, it issues 2 challenges for the same subdomain:

```
Nov 28 21:48:39 localhost caddy[26785]: {"level":"info","ts":1606600119.6782331,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"*.aoalmeida.com","challenge_type":"dns-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Nov 28 21:50:20 localhost caddy[26785]: {"level":"info","ts":1606600220.7814963,"logger":"tls.issuance.acme.acme_client","msg":"validations succeeded; finalizing order","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16420384/193483835"}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.2520313,"logger":"tls.issuance.acme.acme_client","msg":"successfully downloaded available certificate chains","count":2,"first_url":"https://acme-staging-v02.api.letsencrypt.org/acme/cert/fa28c5dadb293c589c96f11af1de9ef671b7"}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.253891,"logger":"tls.issuance.acme","msg":"waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.254234,"logger":"tls.issuance.acme","msg":"done waiting on internal rate limiter","identifiers":["*.aoalmeida.com"]}
Nov 28 21:50:21 localhost caddy[26785]: {"level":"info","ts":1606600221.736078,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"*.aoalmeida.com","challenge_type":"dns-01","ca":"https://acme-v02.api.letsencrypt.org/directory"}
Nov 28 21:52:22 localhost caddy[26785]: {"level":"info","ts":1606600342.8160934,"logger":"tls.issuance.acme.acme_client","msg":"validations succeeded; finalizing order","order":"https://acme-v02.api.letsencrypt.org/acme/order/101130370/6460658836"}
Nov 28 21:52:23 localhost caddy[26785]: {"level":"info","ts":1606600343.3013282,"logger":"tls.issuance.acme.acme_client","msg":"successfully downloaded available certificate chains","count":2,"first_url":"https://acme-v02.api.letsencrypt.org/acme/cert/03194c90cfbc26da8dfff28acdab340aabfb"}
Nov 28 21:52:23 localhost caddy[26785]: {"level":"info","ts":1606600343.302797,"logger":"tls.obtain","msg":"certificate obtained successfully","identifier":"*.aoalmeida.com"}
```

**Comment by mholt** [comment]
> EDIT: as you can see, there is no log message certificate obtained sucessfully with identifier aoalmeida.com. It only obtained one certificate.

Do you already have a certificate for `aoalmeida.com`? If so, that's definitely why.

As for the timeout error, it's probably getting into either this code path:

https://github.com/caddyserver/certmagic/blob/830df5651cba3829aa8d141fcf7a2a93c3d05887/dnsutil.go#L249-L254

Or this one:

https://github.com/caddyserver/certmagic/blob/830df5651cba3829aa8d141fcf7a2a93c3d05887/dnsutil.go#L269-L271

**Comment by Xinayder** [comment]
> Do you already have a certificate for `aoalmeida.com`? If so, that's definitely why.
> 

Yes, I already have.

> As for the timeout error, it's probably getting into either this code path:
> 
> https://github.com/caddyserver/certmagic/blob/830df5651cba3829aa8d141fcf7a2a93c3d05887/dnsutil.go#L249-L254
> 
> Or this one:
> 
> https://github.com/caddyserver/certmagic/blob/830df5651cba3829aa8d141fcf7a2a93c3d05887/dnsutil.go#L269-L271

Should I be able to see the TXT record via `dig` or a similar tool? Also, FWIW Vultr's minimum TTL is 120 seconds.

**Comment by mholt** [comment]
> Also, FWIW Vultr's minimum TTL is 120 seconds.

Ah... CertMagic's default timeout is... 120 seconds. :) Maybe I need to extend it to 5 minutes? (FWIW, you can configure this timeout in Caddy, but I just realized the infamous doc cache bug has affected its documentation page: https://caddyserver.com/docs/json/apps/tls/automation/policies/issuer/acme/challenges/dns/ - there should be a property there called "propagation_timeout" that lets you set this value manually.) Would 5 minutes be a better default you think?

> Should I be able to see the TXT record via dig or a similar tool?

Only briefly, but I think so.

**Comment by Xinayder** [comment]
>  - there should be a property there called "propagation_timeout" that lets you set this value manually.

I haven't seen it before, will try it later this week.

>  Would 5 minutes be a better default you think?

I don't think so. For example, it took me 7 minutes (with the current 2 minute propagation timeout) for my services to start running (I have a reverse proxy and a static website being served by Caddy). Maybe 3 minutes, but it'd be interesting to check the average propagation timeout by common registrars (like Namecheap, gandi.net, etc).

**Comment by mholt** [comment]
Hmm so in that case maybe this is fine, and the built-in retries (CertMagic will retry after 1 minute, then 2, then 2, then 5, then 10, etc) are doing their job just as expected.

**Pull request closed without merging**